internal/venice-client-common/src/main/java/com/linkedin/venice/compression/GzipCompressor.java

```diff
@@ -96,16 +97,22 @@ public ByteBuffer decompress(byte[] data, int offset, int length) throws IOExcep
   }
 
   @Override
-  public ByteBuffer decompressAndPrependSchemaHeader(byte[] data, int length) throws IOException {
-    int valueSchema = ByteUtils.readInt(data, 0);
+  public ByteBuffer decompressAndMaybePrependSchemaHeader(
+      byte[] data,
+      int offset,
+      int length,
+      boolean prependSchemaHeader) throws IOException {
+    if (prependSchemaHeader && offset < SCHEMA_HEADER_LENGTH) {
+      throw new VeniceException("Start offset does not have enough room for schema header.");
+    }
+    int schemaHeader = prependSchemaHeader ? ByteUtils.readInt(data, offset - SCHEMA_HEADER_LENGTH) : 0;
```

Comment:
What is the reason that we can safely assume there will be a valid schema id before the data when the offset is larger than `SCHEMA_HEADER_LENGTH`?

Reply:
Hmm, do you mean we should throw when the offset != SCHEMA_HEADER_LENGTH? Are you suggesting this must be a regression during fetching/decompression logic?

Reply:
The previous logic doesn't assume there is a valid schema id before the data and this PR changed this assumption.
Previously, this function just leaves 4 bytes before the decompressed data in the returned ByteBuffer, so that the user of this function can write the schema id into it to avoid another data copy.

Reply:
Synced offline, we agreed to use a more straightforward API signature to make sure no confusion.